pkg/providers/aws/provider_redissnapshot.go

```diff
@@ -0,0 +1,257 @@
+	// add finalizer to the snapshot cr
+	if err := resources.CreateFinalizer(ctx, p.client, snapshot, DefaultFinalizer); err != nil {
+		errMsg := fmt.Sprintf("failed to set finalizer for snapshot %s", snapshot.Name)
+		return nil, croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	session, err := p.getAwsSession(ctx, redis)
+
+	if err != nil {
+		errMsg := "failed to create AWS session"
+		return nil, croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	cacheSvc := elasticache.New(session)
+
+	return p.createRedisSnapshot(ctx, snapshot, redis, cacheSvc)
+}
+
+func (p *RedisSnapshotProvider) createRedisSnapshot(ctx context.Context, snapshot *v1alpha1.RedisSnapshot, redis *v1alpha1.Redis, cacheSvc elasticacheiface.ElastiCacheAPI) (*providers.RedisSnapshotInstance, croType.StatusMessage, error) {
+	// generate snapshot name
+	snapshotName, err := BuildTimestampedInfraNameFromObjectCreation(ctx, p.client, snapshot.ObjectMeta, DefaultAwsIdentifierLength)
+	if err != nil {
+		errMsg := "failed to generate snapshot name"
+		return nil, croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	// update cr with snapshot name
+	snapshot.Status.SnapshotID = snapshotName
+
+	if err = p.client.Status().Update(ctx, snapshot); err != nil {
+		errMsg := fmt.Sprintf("failed to update instance %s in namespace %s", snapshot.Name, snapshot.Namespace)
+		return nil, croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	// generate cache cluster name
+	clusterName, err := BuildInfraNameFromObject(ctx, p.client, redis.ObjectMeta, DefaultAwsIdentifierLength)
+	if err != nil {
+		errMsg := "failed to get cluster name"
+		return nil, croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	foundSnapshot, err := p.findSnapshotInstance(cacheSvc, snapshotName)
+
+	if err != nil {
+		errMsg := "failed to describe snaphots in AWS"
+		return nil, croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	// get replication group
+	cacheOutput, err := cacheSvc.DescribeReplicationGroups(&elasticache.DescribeReplicationGroupsInput{
+		ReplicationGroupId: aws.String(clusterName),
+	})
+
+	if cacheOutput == nil {
+		errMsg := "snapshot failed, no replication group found"
+		return nil, croType.StatusMessage(errMsg), errorUtil.New(errMsg)
+	}
+
+	// ensure replication group is available
+	if *cacheOutput.ReplicationGroups[0].Status != "available" {
+		errMsg := fmt.Sprintf("current replication group status is %s", *cacheOutput.ReplicationGroups[0].Status)
+		return nil, croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	// find primary cache node
+	var cacheName string
+	for _, i := range cacheOutput.ReplicationGroups[0].NodeGroups[0].NodeGroupMembers {
+		if *i.CurrentRole == "primary" {
+			cacheName = *i.CacheClusterId
+			break
+		}
+	}
```

Comment:
Question: this is copy pasted from the previous redis controller and I didn't make any changes. I didn't check yet but it looks like there's the potential for nil pointer errors here if AWS returned back an empty list. Should I add some more checks?

Reply:
If we add an err handler here https://github.com/integr8ly/cloud-resource-operator/pull/155/files#diff-c5a0bd9058dfdae2fb6c150c7a880efcR108 when you try to describe a resource by `ReplicationGroupId` if it does not find it, it will return an error. So we wont get down this far. 
If it does find it, it will be the first item in the list of `ReplicationGroups`